src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+        val bestBarcodeMetric: SampleBarcodeMetric = this.barcodeToMetrics.getOrElse(sampleBarcode.concatenatedBarcode, unreachable("No metric for sample"))
+        bestBarcodeMetric.reads += 1
+        bestBarcodeMetric.pf_reads += 1
+        if (bestMismatches == 0) {
+          bestBarcodeMetric.perfect_matches += 1
+          bestBarcodeMetric.pf_perfect_matches += 1
+        }
+        else if (bestMismatches == 1) {
+          bestBarcodeMetric.one_mismatch_matches += 1
+          bestBarcodeMetric.pf_one_mismatch_matches += 1
+        }
+        Some(bestSampleOrdinal)
+      }
+    }
+
+    /** Creates a SAM record (always paired) */
+    private def makeSAMRecord(header: SAMFileHeader, fastqRecord: FastqRecord, readStructure: ReadStructure,
+                              readGroupId: String, pairedEnd: Boolean, firstOfPair: Boolean): SAMRecord = {
+      val (readbases, readQualities) = readStructure.structureReadWithQualities(
+        bases     = fastqRecord.bases,
+        qualities = fastqRecord.quals,
+        strict=false
+      ).map { case SubRead(b, q, s) => (b, q.getOrElse(unreachable())) }.unzip
+      val record: SAMRecord = new SAMRecord(header)
+      val readBases = readbases.mkString.getBytes
+      val baseQualities = readQualities.mkString.getBytes
+
+      record.setReadName(fastqRecord.name)
+      record.setReadBases(readBases)
+      record.setReadUnmappedFlag(true)
+      if (pairedEnd) {
+        record.setReadPairedFlag(true)
+        record.setMateUnmappedFlag(true)
+        if (firstOfPair) record.setFirstOfPairFlag(true)
+        else record.setSecondOfPairFlag(true)
+      }
+      record.setAttribute(ReservedTagConstants.READ_GROUP_ID, readGroupId)
+
+      convertQuality(baseQualities, qualityFormat)
+      baseQualities.foreach {
+        qual =>
+          val uQual: Int = qual & 0xff
+          if (uQual < minQ || uQual > maxQ) {
+            throw new IllegalStateException(s"Base quality $uQual is not in the range $minQ ... $maxQ for read ${fastqRecord.header}")
+          }
+      }
+      record.setBaseQualities(baseQualities)
+
+      record
+    }
+
+    /** Based on the type of quality scores coming in, converts them to a numeric byte[] in phred scale. */
+    private[fastq] def convertQuality(quals: Array[Byte], version: FastqQualityFormat) {
+      version match {
+        case FastqQualityFormat.Standard => SAMUtils.fastqToPhred(quals)
+        case FastqQualityFormat.Solexa => solexaQualityConverter.convertSolexaQualityCharsToPhredBinary(quals)
+        case FastqQualityFormat.Illumina => solexaQualityConverter.convertSolexa_1_3_QualityCharsToPhredBinary(quals)
+      }
+    }
+
+    /** Gets the sample barcode bases */
+    private def getSampleBarcode(records: Seq[FastqRecord]): Seq[String] = {
+      records.zip(sampleBarcodeReadStructures).map {
+        case (rec, rs) => rs.structureRead(rec.bases).map(_.bases).mkString
```

Comment:
Two things (applies to `getMolecularBarcodes` too:
1) I think you should filter out empty values here - it's weird to return empties
2) I think this will do the wrong thing _if_ there are two barcodes of the same type in the same read.  Imagine a read structure of `5M8S5M` - the two molecular barcodes (one each side of the sample barcode) will get concatenated and returned as a single 10-base string instead of being concatenated with the delimiter between.